src/extension/src/coreTemplateStudio.ts

```diff
@@ -94,144 +84,151 @@ export class CoreTemplateStudio {
     }
   }
 
-  private constructor(process: ChildProcess, port: number, url: string) {
-    this._process = process;
-    this._port = port;
-    this._url = url;
+  private constructor(processCli: ChildProcess) {
+    this._processCli = processCli;
+    this.cliEvents = new CliEventEmitter();
+    this.readStream(this._processCli);
   }
 
-  public getPort(): number {
-    return this._port;
+  // This function is a listener, in the constructor, it gets attached
+  // then it will always get triggered when a command is write to the cli and there is responses from cli, until the process gets killed
+  public async readStream(process: ChildProcess) {
+    let data = "";
+    process.stdout.on("data", (chunk) => {
+      data += chunk;
+      var responses = data.toString().split("\n");
+      for (let i = 0; i < responses.length - 1; i++) {
+        var result = JSON.parse(responses[i]);
+        this.cliEvents.emit(result['messageType'], result['content']);
+      }
+      data = responses[responses.length - 1];
+    });
+
+    process.stderr.on("data", (data) => {
+      this.cliEvents.emit("error", data.toString());
+    });
   }
 
-  public async getProjectTypes(): Promise<any> {
-    // TODO: use this in client instead of fetching directly from API server
-    const url = new URL(CONSTANTS.API.ENDPOINTS.PROJECT_TYPE, this._url);
-    return await fetch(url.href, { method: CONSTANTS.API.METHODS.GET })
-      .then((response: Response) => {
-        return response.json();
-      })
-      .catch((error: Error) => {
-        throw Error(error.toString());
+  private async awaitCliEvent(eventName: string): Promise<any> {
+    return new Promise((resolve, reject) => {
+      this.cliEvents.once(eventName, (data) => {
+        this.cliEvents.removeAllListeners();
+        resolve(data);
+      });
+      this.cliEvents.once("error", (data) => {
```

Comment:
Suppose your have two events `X` and `Y` and you subscribe to both of them. If any of these two operation fails and sends the `error` event, all listeners to both events will be removed even though one of them did not actually fail.

Is this the intended behaviour? If not, maybe for each event `E`, you should have an additional error with `-error` (or anything like that) appended to its name.

Reply:
An additional note on style, `EventEmitter.once` return `this`, so you can chain calls to once like this: `this.cliEvents.once(...).once(...)`

Reply:
> Is this the intended behavior? From what I read from the older code, for each operation, there is an associated failure (instead of a global one).
Since the CLI works synchronously, there is only one command running at a time (so only one event other than error being subscribed). I think because of this, it should work.

Reply:
Thank you for pointing out the potential concerns and style suggestion. As Juan explained, because the way the Cli setup, we are good on the case you pointed. I updated my PR with your style suggestion 👍